### ISSUE TITLE
Data passed to template does not update per action in array

### ISSUE THREAD

**Issue by aminimalanimal** [amwmedia/plop#145]
I'm prompting the user to write a comma-separated list of components to generate, parsing them into an array, then pushing to the actions array with the name of the component specified in the data parameter. All of these components use the same template, and I'm relying on the data I'm changing to differentiate them. Unfortunately, the data doesn't seem to be updated when it is passed to the next item in the `actions` array.

```
    ...
    console.log(subcomponents) // [ 'MyComponentFoo', 'MyComponentBar' ]

    if (subcomponents) {
      subcomponents.forEach((subcomponent) => {
        actions.push({
          type: 'add',
          path: `../../components/{{ properCase name }}/${subcomponent}.vue`,
          templateFile: './component/subcomponent.vue.hbs',
          abortOnFail: true,
          data: {
            subcomponent,
          },
        })
      })
    }
```

When logging my `actions` array, I see the expected `data` is placed in these items:
```
  {
    "type": "add",
    "path": "../../components/{{ properCase name }}/MyComponentFoo.vue",
    "templateFile": "./component/subcomponent.vue.hbs",
    "abortOnFail": true,
    "data": {
      "subcomponent": "MyComponentFoo"
    }
  },
  {
    "type": "add",
    "path": "../../components/{{ properCase name }}/MyComponentBar.vue",
    "templateFile": "./component/subcomponent.vue.hbs",
    "abortOnFail": true,
    "data": {
      "subcomponent": "MyComponentBar"
    }
  }
```

In my subcomponent.vue.hbs template, I'm referencing the `subcomponent` data I'm passing:
```
{{ subcomponent }}
```

Since I used a template literal\* to create the `path`, both files are generated, but both files end up with `MyComponentFoo` as their contents. It seems that subsequent `data` specifications don't override earlier ones.

\*(Attempting to use `{{ subcomponent }}` as a handlebars template in the `path` instead of using a template literal failed because it attempted to create `MyComponentFoo.vue` twice.)

**Comment by amwmedia** [comment]
yeah, I agree, this is confusing. The `data` attribute should override, but currently it can only add new props. I'm going to mark this as a bug. As a workaround you should be able to use a function in place of your data object and set the value on the answer object directly inside that function.

example:
```javascript
const {subcomponents} = answerData;
if (subcomponents) {
  subcomponents.forEach((subcomponent) => {
    actions.push({
      type: 'add',
      path: `../../components/{{ properCase name }}/{{subcomponent}}.vue`,
      templateFile: './component/subcomponent.vue.hbs',
      abortOnFail: true,
      data: () => {
        answerData.subcomponent = subcomponent;
        return {};
      },
    })
  })
}
```

**Comment by amwmedia** [comment]
I've been looking at this due to @apurvaojas posting this issue on [node-plop as well](https://github.com/amwmedia/node-plop/issues/89) and I have 2 possible solutions that I'd like to get feedback on.

1. Attach `data` to the answer data under a `.data` property instead of mixing the data in.
2. Mix `data` into the main answer data and always override anything that overlaps.

both of these options break the existing contract, so we would probably need to increment the major version number (which is fine, but something to note).

**Comment by amwmedia** [comment]
the more I'm thinking about this, the more I'm leaning toward a `data` property that can be accessed by the template.

**Comment by aminimalanimal** [comment]
Both of them seem like decent solutions.

Some brainstorming:
- I wonder if #1 would cause the need to read or write to `data.data` in the `data` method of an `actions` object. The current documentation does name the answer-data `data` when you pass it into the `actions` method, and if someone were to do something like you showed me above, you might end up in that somewhat awkward position.
- I'm assuming based on the way you describe it that #1 would require the template to have things like `{{#if data.subcomponents}}` instead of `{{#if subcomponents}}`. If that's the case, #1 has another drawback: it forces the template to care about whether its data was constructed directly by inquirer prompts or derived afterward. Needing to care for that kind of discrepancy could potentially limit reuse (or impose certain coding patterns that enable reuse, like always explicitly assigning answer data to the data object and always using `data.thing` in the templates).

#2 sounds like how I assumed it'd work. My assumption is that it'd override _per object that declares the override_ and reset back to its original value for any object in the actions array that doesn't explicitly override it. Personally, I think I prefer this method as it seems cleanest for the templates.

**Comment by amwmedia** [comment]
While reading your brainstorming points above, I came up with an option 3 that checks all the boxes. We can add some logic to "clean up" any attributes that were added by the `data` object after the action completes. I actually don't think this even breaks the existing contract because don't believe that there was any expectation that the values provided in the `data` prop should persist to future actions. If anything, the fact that they persisted was the defect.

**Comment by apurvaojas** [comment]
There is a [test case](https://github.com/amwmedia/node-plop/blob/master/tests/custom-data-in-actions.ava.js#L24) in `node-plop` which expect `data` prop to persist in future actions. 

If this is a bug then [this solution](https://github.com/amwmedia/node-plop/pull/102#issuecomment-466739364) can work.

**Comment by amwmedia** [comment]
The solution I came up with does not break this test, I'll explain through an example.

``` javascript
plop.setGenerator('test', {
  prompts: [{
    type: 'input',
    name: 'fromInquirer'
  }],
  actions: [
    {
      type: 'add',
      data: {
        fromInquirer: 'not applied',
        fromData: 'A'
      }
    },
    data => data.mutated = 'true',
    {
      type: 'add',
      data: {
        mutated: 'not applied',
        fromData: 'B'
      }
    }
  ]
});
```
In the example above:
- `data.fromInquirer` in the action 1 does not get applied because the action data properties are not allowed to overwrite prompt data.
- `data.fromData` is applied and then cleaned up after the action completes.
- action 2 is a custom action function that mutates the main data object, these changes are persisted to action 3.
- `data.mutated` in action 3 is not applied because `mutated` from action 2 is treated the same as prompt data.
- `data.fromData` in action 3 is applied and then cleaned up after the action completes.

clear as mud?

**Comment by aminimalanimal** [comment]
Interesting. I'd never considered mixing functions that manipulate data with the action objects.

Would this overwrite the inquirer prompt-provided data?
```
plop.setGenerator('test', {
  prompts: [{
    type: 'input',
    name: 'fromInquirer'
  }],
  actions: [
    data => data.fromInquirer = 'true',
    {
      type: 'add',
      data: {
        fromInquirer: 'false',
      }
    }
  ]
});
```

**Comment by amwmedia** [comment]
The function would, the data array would not. fromInquirer would be "true" during the second action (add).

**Comment by amwmedia** [comment]
fixed in node-plop v0.18.0, this change will make its way into the next version of plop soon

**Comment by amwmedia** [comment]
fixed in v2.3.0, now available on npm
